cpp/include/cudf/join.hpp

```diff
@@ -647,5 +649,206 @@ class hash_join {
   const std::unique_ptr<const hash_join_impl> impl;
 };
 
+/**
+ * @brief Returns a pair of row index vectors corresponding to all pairs
+ * of rows between the specified tables where the predicate evaluates to true.
+ *
+ * The first returned vector contains the row indices from the left
+ * table that have a match in the right table (in unspecified order).
+ * The corresponding values in the second returned vector are
+ * the matched row indices from the right table.
+ *
+ * @code{.pseudo}
+ * Left: {{0, 1, 2}}
+ * Right: {{1, 2, 3}}
+ * Expression: Left.Column_0 == Right.Column_0
+ * Result: {{1, 2}, {0, 1}}
+ *
+ * Left: {{0, 1, 2}, {3, 4, 5}}
+ * Right: {{1, 2, 3}, {4, 6, 7}}
+ * Expression: (Left.Column_0 == Right.Column_0) AND (Left.Column_1 == Right.Column_1)
+ * Result: {{1}, {0}}
+ * @endcode
+ *
+ * @throw cudf::logic_error if number of elements in `left_keys` or `right_keys`
+ * mismatch.
+ *
+ * @param left The left table
+ * @param right The right table
+ * @param binary_predicate The condition on which to join.
+ * @param compare_nulls Whether the equality operator returns true or false for two nulls.
+ * @param mr Device memory resource used to allocate the returned table and columns' device memory
+ *
+ * @return A pair of vectors [`left_indices`, `right_indices`] that can be used to construct
+ * the result of performing a conditional inner join between two tables `left` and `right` .
+ */
+std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+          std::unique_ptr<rmm::device_uvector<size_type>>>
+conditional_inner_join(
+  table_view left,
+  table_view right,
+  ast::expression binary_predicate,
+  null_equality compare_nulls         = null_equality::EQUAL,
```

Comment:
Since the caller can specify the join condition via the AST, I'm surprised to see this parameter here.  Instead I would expect there to be either two types of AST equality expressions or a parameterized AST equality operator to control the null behavior.  The caller would then build the appropriate type of AST to match the null handling they need.

Reply:
Yeah, we were just talking about this earlier today. It quickly devolved into asking whether or not we need two versions of every operator that has different null behavior. We kept the `null_equality` for now to be consistent with the other join APIs, but I agree it is weird to have it and we should do something better.

Reply:
@jrhemstad and I had a similar discussion earlier today, and I think we decided moving in that direction makes sense as well.

Reply:
I think the only operators for which we might want this are `=` and `!=`, right? I can't imagine reasonable semantics for any other operator. `!=` is borderline, I can imagine that other tools might try to define different semantics for `NULL != NULL` but I don't know what those are for consumers of this API.

Reply:
I suspect it will only be necessary for `==`.  That being said, in the future it could be very useful in an AST expression to conditionally branch based on the result of an "is this null" operator.  Then users could implement their own fancy-pants versions of null comparison or other special-case null handling in their custom expressions.